src/mysky/encrypted_files.ts

```diff
@@ -0,0 +1,330 @@
+import {
+  throwValidationError,
+  validateBoolean,
+  validateHexString,
+  validateNumber,
+  validateObject,
+  validateString,
+  validateUint8Array,
+  validateUint8ArrayLen,
+} from "../utils/validation";
+
+export const ENCRYPTED_JSON_RESPONSE_VERSION = 1;
+
+/**
+ * The length of the encryption key.
+ */
+export const ENCRYPTION_KEY_LENGTH = 32;
+
+/**
+ * The length of the hidden field metadata stored along with encrypted files.
+ * Note that this is hidden from the user, but not actually encrypted. It is
+ * stored after the nonce.
+ */
+export const ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH = 16;
+
+/**
+ * The length of the random nonce, prepended to the encrypted bytes. It comes
+ * before the unencrypted hidden field metadata.
+ */
+export const ENCRYPTION_NONCE_LENGTH = 24;
+
+/**
+ * The length of the overhead introduced by encryption.
+ */
+const ENCRYPTION_OVERHEAD_LENGTH = 16;
+
+/**
+ * The length of the share-able path seed.
+ */
+export const ENCRYPTION_PATH_SEED_LENGTH = 32;
+
+export type EncryptedJSONResponse = {
+  data: JsonData | null;
+};
+
+type DerivationPathObject = {
+  pathSeed: Uint8Array;
+  directory: boolean;
+  name: string;
+};
+
+type EncryptedFileMetadata = {
+  // 8-bit uint.
+  version: number;
+};
+
+/**
+ * Decrypts the given bytes as an encrypted JSON file.
+ *
+ * @param data - The given raw bytes.
+ * @param key - The encryption key.
+ * @returns - The JSON data and metadata.
+ * @throws - Will throw if the bytes could not be decrypted.
+ */
+export function decryptJSONFile(data: Uint8Array, key: Uint8Array): JsonFullData {
+  validateUint8Array("data", data, "parameter");
+  validateUint8ArrayLen("key", key, "parameter", ENCRYPTION_KEY_LENGTH);
+
+  // Validate that the size of the data corresponds to a padded block.
+  if (!checkPaddedBlock(data.length)) {
+    throwValidationError("data", data, "parameter", `padded encrypted data, length was '${data.length}'`);
+  }
+
+  // Extract the nonce.
+  const nonce = data.slice(0, ENCRYPTION_NONCE_LENGTH);
+  data = data.slice(ENCRYPTION_NONCE_LENGTH);
+
+  // Extract the unencrypted hidden field metadata.
+  const metadataBytes = data.slice(0, ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
+  data = data.slice(ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
+  const metadata = decodeEncryptedFileMetadata(metadataBytes);
+  if (metadata.version !== ENCRYPTED_JSON_RESPONSE_VERSION) {
+    throw new Error(
+      `Received unrecognized JSON response version '${metadata.version}', expected '${ENCRYPTED_JSON_RESPONSE_VERSION}'`
+    );
+  }
+
+  // Decrypt the non-nonce part of the data.
+  let decryptedBytes = secretbox.open(data, nonce, key);
+  if (!decryptedBytes) {
+    throw new Error("Could not decrypt given encrypted JSON file");
+  }
+
+  // Trim the 0-byte padding off the end.
+  let paddingIndex = decryptedBytes.length;
+  while (paddingIndex > 0 && decryptedBytes[paddingIndex - 1] === 0) {
+    paddingIndex--;
+  }
```

Comment:
is there a possibility that you will remove a real data or real data is never zero at the end ? what type of data is `decryptedBytes` ? uint[] or a string ?

Reply:
There should never be data loss because JSON must end with a '}'. Added a comment.